### PR TITLE
Adding StreamingCompletion message (C#)

### PR DESCRIPTION
**Pull request by moozzyk**

### CODE REVIEW

**Comment by moozzyk** [comment]
Soliciting early feedback. TS part to be added.

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/CompletionMessage.cs [hunk 1]
```diff
@@ -10,31 +10,50 @@ public class CompletionMessage : HubMessage
         public string Error { get; }
         public object Result { get; }
         public bool HasResult { get; }
+        public bool IsStreamingCompletion { get; }
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/CompletionMessage.cs]: The question is whether to use this flag or create a separate `StreamCompletion` message which will not have Result and `HasResult` properties (possibly a base completion class)

Comment (BrennanConroy) [reply src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/CompletionMessage.cs]: I think it should be a new class

Comment (moozzyk) [reply src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/CompletionMessage.cs]: I am adding a new class in TypeScript client to how it looks like.

File: client-ts/Microsoft.AspNetCore.SignalR.Client.TS/HubConnection.ts [hunk 2]
```diff
@@ -63,9 +63,10 @@ export class HubConnection {
                     break;
                 case MessageType.Result:
                 case MessageType.Completion:
+                case MessageType.StreamCompletion:
                     let callback = this.callbacks.get(message.invocationId);
                     if (callback != null) {
-                        if (message.type == MessageType.Completion) {
+                        if (message.type == MessageType.Completion || message.type == MessageType.StreamCompletion) {
```
Comment (muratg) [inline client-ts/Microsoft.AspNetCore.SignalR.Client.TS/HubConnection.ts]: `===` instead of `==`?

File: src/Microsoft.AspNetCore.SignalR.Client.Core/InvocationRequest.cs [hunk 3]
```diff
@@ -133,10 +145,24 @@ public NonStreaming(CancellationToken cancellationToken, Type resultType, string
...
+                if (completionMessage.IsStreamingCompletion)
+                {
+                    Logger.ReceivedUnexpectedStreamingCompletion(InvocationId);
+                    _completionSource.TrySetException(new InvalidOperationException("Server returned streaming completion message for a non-streamed invocation."));
+                    return;
+                }
+
+                if (!string.IsNullOrEmpty(completionMessage.Error))
+                {
+                    Fail(new HubException(completionMessage.Error));
```
Comment (muratg) [inline src/Microsoft.AspNetCore.SignalR.Client.Core/InvocationRequest.cs]: Should the error message be some non-empty, descriptive string instead?

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/CompletionMessage.cs [hunk 4]
```diff
@@ -10,31 +10,50 @@ public class CompletionMessage : HubMessage
...
+
             Error = error;
             Result = result;
             HasResult = hasResult;
+            IsStreamingCompletion = isStreamingCompletion;
         }
 
         public override string ToString()
         {
+            var caption = $"{(IsStreamingCompletion ? "Streaming" : string.Empty)} Completion";
```
Comment (muratg) [inline src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/CompletionMessage.cs]: nit: `$"{(IsStreamingCompletion ? "Streaming " : string.Empty)}Completion"`

File: src/Microsoft.AspNetCore.SignalR.Client.Core/InvocationRequest.cs (hunk 3 above)
Comment (moozzyk) [reply src/Microsoft.AspNetCore.SignalR.Client.Core/InvocationRequest.cs]: It is. This is the error thrown on the server side.

File: test/Microsoft.AspNetCore.SignalR.Common.Tests/Internal/Protocol/MessagePackHubProtocolTests.cs [hunk 5]
```diff
@@ -35,6 +35,9 @@ public class MessagePackHubProtocolTests
             new object[] { new[] { new CompletionMessage("xyz", error: null, result: new CustomObject(), hasResult: true) } },
             new object[] { new[] { new CompletionMessage("xyz", error: null, result: new[] { new CustomObject(), new CustomObject() }, hasResult: true) } },
 
+            new object[] { new[] { new StreamCompletionMessage("xyz", error: null) } },
```
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.Common.Tests/Internal/Protocol/MessagePackHubProtocolTests.cs]: @BrennanConroy  - I just noticed you did not add tests for CancellationMessage here

File: client-ts/Microsoft.AspNetCore.SignalR.Client.TS/MessagePackHubProtocol.ts [hunk 6]
```diff
@@ -97,7 +99,36 @@ export class MessagePackHubProtocol implements IHubProtocol {
...
+        }
+
+        const errorResult = 1;
+        const voidResult = 2;
+        const nonVoidResult = 3;
+
+        let resultKind = properties[2];
+
+        /*
+        if ((resultKind === nonVoidResult)
```
Comment (davidfowl) [inline client-ts/Microsoft.AspNetCore.SignalR.Client.TS/MessagePackHubProtocol.ts]: ?

File: src/Microsoft.AspNetCore.SignalR.Client.Core/Internal/SignalRClientLoggerExtensions.cs [hunk 7]
```diff
@@ -40,44 +40,53 @@ internal static class SignalRClientLoggerExtensions
         private static readonly Action<ILogger, string, Exception> _dropStreamMessage =
             LoggerMessage.Define<string>(LogLevel.Warning, new EventId(9, nameof(DropStreamMessage)), "Dropped unsolicited StreamItem message for invocation '{invocationId}'.");
 
+        private static readonly Action<ILogger, string, Exception> _dropStreamCompletionMessage =
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Client.Core/Internal/SignalRClientLoggerExtensions.cs]: This is always a bad idea. You can't add logging events in between. Just add it at the end. We're going to have to do that once we ship anyways.

File: test/Microsoft.AspNetCore.SignalR.Common.Tests/Internal/Protocol/MessagePackHubProtocolTests.cs (hunk 5 above)
Comment (moozzyk) [reply test/Microsoft.AspNetCore.SignalR.Common.Tests/Internal/Protocol/MessagePackHubProtocolTests.cs]: Nevermind - I added it.

File: client-ts/Microsoft.AspNetCore.SignalR.Client.TS/MessagePackHubProtocol.ts (hunk 6 above)
Comment (moozzyk) [reply client-ts/Microsoft.AspNetCore.SignalR.Client.TS/MessagePackHubProtocol.ts]: TS is WIP

File: src/Microsoft.AspNetCore.SignalR.Client.Core/Internal/SignalRClientLoggerExtensions.cs (hunk 7 above)
Comment (moozzyk) [reply src/Microsoft.AspNetCore.SignalR.Client.Core/Internal/SignalRClientLoggerExtensions.cs]: OK - I thought it was fine because we did not ship anything yet.. . How about deleting messages? Should we leave the gap?

Comment (davidfowl) [reply src/Microsoft.AspNetCore.SignalR.Client.Core/Internal/SignalRClientLoggerExtensions.cs]: I'm ok fixing it now (you already did it), we shouldn't do it after we ship.

File: client-ts/Microsoft.AspNetCore.SignalR.Client.TS/HubConnection.ts (hunk 2 above)
Comment (moozzyk) [reply client-ts/Microsoft.AspNetCore.SignalR.Client.TS/HubConnection.ts]: Actually, this is inside switch so it does not matter.

**Comment by moozzyk** [comment]
The end to end JS client test fail and I see a lot of failures like this:
```
warn: Microsoft.AspNetCore.Server.Kestrel[0]
      Connection processing ended abnormally.
System.InvalidOperationException: Cursor is out of bounds
   at System.IO.Pipelines.ReadCursor.ThrowOutOfBoundsException()
   at System.IO.Pipelines.ReadableBuffer.Slice(ReadCursor start, ReadCursor end)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.ParseRequest(ReadableBuffer buffer, ReadCur
sor& consumed, ReadCursor& examined)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.<ParseRequestAsync>d__37.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.<ProcessRequestsAsync>d__186.MoveNext()
info: Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets[14]
```
@halter73 - does this ring a bell? I have not fully investigated this but I don't remember seeing this kind of warnings before. Also, when running tests one by one they pass.

File: test/Microsoft.AspNetCore.SignalR.Client.Tests/HubConnectionProtocolTests.cs [hunk 8]
```diff
@@ -266,7 +293,57 @@ public class HubConnectionProtocolTests
...
+                Assert.Equal("Streaming methods must be invoked using StreamAsync.", ex.Message);
+            }
+            finally
+            {
+                await hubConnection.DisposeAsync().OrTimeout();
+                await connection.DisposeAsync().OrTimeout();
+            }
+        }
+
+        [Fact]
```
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.Client.Tests/HubConnectionProtocolTests.cs]: Note to myself - need to add a counterpart of this for the Stream method as well as a test case for invoking a streaming method returning an empty stream with Invoke

**Comment by halter73** [comment]
@moozzyk It looks like you're running into https://github.com/dotnet/corefxlab/issues/1855

**Comment by moozzyk** [comment]
Ready to review.

**Comment by halter73** [comment]
Don't take my word for it, but it LGTM 👍

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/MessagePackHubProtocol.cs [hunk 9]
```diff
@@ -132,6 +136,16 @@ private static CompletionMessage CreateCompletionMessage(Unpacker unpacker, IInv
             return new CompletionMessage(invocationId, error, result, hasResult);
         }
 
+        private static StreamCompletionMessage CreateStreamCompletionMessage(Unpacker unpacker, long arraySize, IInvocationBinder binder)
+        {
+            Debug.Assert(arraySize == 2 || arraySize == 3, "Unexpected item count");
+
+            var invocationId = ReadInvocationId(unpacker);
+            // Error is optional so StreamCompletion without error will has 3 items, StreamCompletion with error has 4 items
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/MessagePackHubProtocol.cs]: "will have" and it should be 2 and 3?

File: client-ts/Microsoft.AspNetCore.SignalR.Client.TS/HubConnection.ts [hunk 10]
```diff
@@ -92,14 +93,8 @@ export class HubConnection {
...
     private connectionClosed(error?: Error) {
-        let errorCompletionMessage = <CompletionMessage>{
-            type: MessageType.Completion,
-            invocationId: "-1",
-            error: error ? error.message : "Invocation cancelled due to connection being closed.",
-        };
-
         this.callbacks.forEach(callback => {
-            callback(errorCompletionMessage);
+            callback(undefined, error ? error : new Error("Invocation cancelled due to connection being closed."));
```
[the remark this answers is not in the document]
Comment (moozzyk) [reply client-ts/Microsoft.AspNetCore.SignalR.Client.TS/HubConnection.ts]: @anurse to the rescue!

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/MessagePackHubProtocol.cs (hunk 9 above)
Comment (moozzyk) [reply src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/MessagePackHubProtocol.cs]: should be 2 and 3 - 3 and 4 was a bug which I found by running end to end JS test. Fixed the code but missed the comment

File: client-ts/Microsoft.AspNetCore.SignalR.Client.TS/HubConnection.ts (hunk 2 above)
Comment (anurse) [reply client-ts/Microsoft.AspNetCore.SignalR.Client.TS/HubConnection.ts]: Usage of `===` is generally [always preferred](https://palantir.github.io/tslint/rules/triple-equals/) no matter where it occurs (even if it would be "safe" to use `==`). It's a code style and clarity thing rather than a correctness thing. We should consider running a linter like [TSLint](https://palantir.github.io/tslint/) on our code
